modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp

```diff
@@ -0,0 +1,2000 @@
+    const float *verify_data;
+    float *data;
+    float *tmp_verify_data;
+    ocl::Queue queue = ocl::Queue::getDefault();
+    data = reinterpret_cast<float *>(clEnqueueMapBuffer((cl_command_queue)queue.ptr(),
+                                                        (cl_mem)top, true, CL_MAP_READ,
+                                                        0, sizeof(float) * numImages * top_dim_,
+                                                        0, NULL, NULL, NULL));
+    tmp_verify_data =
+        reinterpret_cast<float *>(clEnqueueMapBuffer((cl_command_queue)queue.ptr(),
+                                                     (cl_mem)verify_blob, true, CL_MAP_READ,
+                                                     0, sizeof(float) * numImages * top_dim_,
+                                                     0, NULL, NULL, NULL));
+    verify_data = tmp_verify_data;
+
+    for (int32_t n = 0; n < numImages; ++n) {
+        for (int32_t g = 0; g < group_; ++g) {
+            int32_t output_image_offset = n * top_dim_ + output_w_ * output_h_ * M_ * g;
+            for (int out_ch = 0; out_ch < M_ && !verificationFail; out_ch++)
+                for (int h = 0; h < output_h_ && !verificationFail; h++)
+                    for (int w = 0; w < output_w_; w++) {
+                        size_t offset = output_image_offset + out_ch * output_w_ * output_h_ + h * output_w_ + w;
+                        if (fabs(data[offset] - verify_data[offset]) > 0.1 * fabs(verify_data[offset]) &&
+                            !(fabs(verify_data[offset]) < 1.e-3 &&
+                            fabs(data[offset] - verify_data[offset]) < 1.e-4))
+                        {
+                            dbgPrint(printf("test verification failed @ image %d group %d"
+                                            "out_ch %d h %d w %d got %G expected %G\n",
+                                            n, g, out_ch, h, w, data[offset], verify_data[offset]));
+                            verificationFail = 1;
+                            goto out;
+                        }
+                    }
+        }
+    }
+
+out:
+    clEnqueueUnmapMemObject((cl_command_queue)queue.ptr(),
+                            (cl_mem)top, data, 0, NULL, NULL);
+    clEnqueueUnmapMemObject((cl_command_queue)queue.ptr(),
+                            (cl_mem)verify_blob, tmp_verify_data, 0, NULL, NULL);
+    if (verificationFail == 1)
+        return false;
+    else
+        return true;
+}
+
+template<typename Dtype>
+void OCL4DNNConvSpatial<Dtype>::unloadProgram(ocl::Program& prog)
+{
+    ocl::Context ctx = ocl::Context::getDefault();
+    ctx.unloadProg(prog);
+}
+
+template<typename Dtype>
+ocl::Program OCL4DNNConvSpatial<Dtype>::compileKernel()
+{
+    String errmsg;
+    ocl::Context ctx = ocl::Context::getDefault();
+    ocl::ProgramSource src(kernel_.c_str());
+    ocl::Program program = ctx.getProg(src, options_, errmsg);
+    CV_Assert(!kernel_name_.empty());
```

Comment:
Program compilation may fail due unsupported option: `-cl-no-subgroup-ifp` (it is described in OpenCL 2.1 only, OpenCV supports 1.2+ and should not fail on OpenCL 1.1).
1. Using of this option should be optional (or detected separately).
2. OpenCL program compilation errors should be handled properly. (It fails in another place with "Basic kernel failed verification" message)

Problem is observed on this configuration:
```
OpenCL Platforms:
    Intel(R) OpenCL
        iGPU: Intel(R) HD Graphics 520 (OpenCL 2.0 )
        CPU: Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz (OpenCL 2.0 (Build 10094))
Current OpenCL device:
    Type = iGPU
    Name = Intel(R) HD Graphics 520
    Version = OpenCL 2.0
    Driver version = 20.19.15.4463
    Compute units = 24
    Max work group size = 256
    Local memory size = 64 kB
    Max memory allocation size = 805 MB 264 kB
    Double support = No
    Host unified memory = Yes
    Has AMD Blas = No
    Has AMD Fft = No
    Preferred vector width char = 1
    Preferred vector width short = 1
    Preferred vector width int = 1
    Preferred vector width long = 1
    Preferred vector width float = 1
    Preferred vector width double = 0
```

Reply:
yes, we have checked this build option based on opencl version and carefully check the program compilation result if it fails.